src/coreComponents/physicsSolvers/fluidFlow/wells/CompositionalMultiphaseWellKernels.hpp

```diff
@@ -750,146 +751,90 @@ struct ResidualNormKernel
-          integer const numComponents,
-          arrayView1d< globalIndex const > const & wellElemDofNumber,
-          arrayView1d< integer const > const & wellElemGhostRank,
-          arrayView1d< real64 const > const & wellElemPressure,
-          arrayView2d< real64 const, compflow::USD_COMP > const & wellElemCompDens,
-          integer const allowCompDensChopping,
-          real64 const scalingFactor )
+  static integer
+  createAndLaunch( integer const allowCompDensChopping,
+                   real64 const scalingFactor,
+                   globalIndex const rankOffset,
+                   integer const numComp,
+                   string const dofKey,
+                   ElementSubRegionBase const & subRegion,
+                   arrayView1d< real64 const > const localSolution )
   {
-    using COFFSET = compositionalMultiphaseWellKernels::ColOffset;
-
-    real64 constexpr eps = minDensForDivision;
-
-    RAJA::ReduceMin< ReducePolicy< POLICY >, localIndex > minVal( 1 );
-
-    forAll< POLICY >( wellElemDofNumber.size(), [=] GEOSX_HOST_DEVICE ( localIndex const iwelem )
-    {
-      if( wellElemGhostRank[iwelem] < 0 )
-      {
-        // pressure
-        localIndex lid = wellElemDofNumber[iwelem] + COFFSET::DPRES - rankOffset;
-        real64 const newPres = wellElemPressure[iwelem] + scalingFactor * localSolution[lid];
-
-        // the pressure must be positive
-        if( newPres < 0.0 )
-        {
-          minVal.min( 0 );
-        }
-
-        // if component density is not allowed, the time step fails if a component density is negative
-        // otherwise, we just check that the total density is positive, and negative component densities
-        // will be chopped (i.e., set to zero) in applySystemSolution
-        if( !allowCompDensChopping )
-        {
-          for( integer ic = 0; ic < numComponents; ++ic )
-          {
-            lid = wellElemDofNumber[iwelem] + ic + 1 - rankOffset;
-            real64 const newDens = wellElemCompDens[iwelem][ic] + scalingFactor * localSolution[lid];
-
-            if( newDens < 0 )
-            {
-              minVal.min( 0 );
-            }
-          }
-        }
-        else
-        {
-          real64 totalDens = 0.0;
-          for( integer ic = 0; ic < numComponents; ++ic )
-          {
-            lid = wellElemDofNumber[iwelem] + ic + 1 - rankOffset;
-            real64 const newDens = wellElemCompDens[iwelem][ic] + scalingFactor * localSolution[lid];
-            totalDens += (newDens > 0.0) ? newDens : 0.0;
-          }
-          if( totalDens < eps )
-          {
-            minVal.min( 0 );
-          }
-        }
-      }
-    } );
-    return minVal.get();
+    arrayView1d< real64 const > const pressure =
+      subRegion.getExtrinsicData< extrinsicMeshData::well::pressure >();
+    arrayView2d< real64 const, compflow::USD_COMP > const compDens =
+      subRegion.getExtrinsicData< extrinsicMeshData::well::globalCompDensity >();
```

Comment:
Sorry, I work on a small screen, so I don't like long lines :) . But I put everything on the same line in the latest commit

Reply:
I also split these as they often break even generous column limits like 160.
IMHO this should be (in some distant and bright future)
```
    auto const compDens = subRegion.getField< fields::well::globalCompDensity >();
```

Reply:
> ```
>     auto const compDens = subRegion.getField< fields::well::globalCompDensity >().toViewConst();
> ```

I'm not really sure about the auto. It's always tedious to swim upstream to find what the type you're manipulating!
But `field` instead of `extrinsicData` is a happy 14 characters reduction. 🎉 (plus I understand).
We could have a `getFieldV/getFieldView` and `getFieldCV/getFieldConstView` for the `views` and `const views`?

> in some distant and bright future

You say this because we won't be there! 🤣 🤣 🤣